tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/**
+ * OAI-PMH harvester factory integration test.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2010.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+namespace VuFindTest\Harvest\OaiPmh;
+
+use Laminas\Http\Client;
+use VuFindHarvest\OaiPmh\Communicator;
+
+/**
+ * OAI-PMH communicator test.
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+class CommunicatorTest extends \PHPUnit\Framework\TestCase
+{
+
+    /**
+     * Get Communicator
+     *
+     * @param string                     $uri       Base URI for OAI-PMH server
+     * @param Client                     $client    HTTP client
+     *
+     * @return type
+     */
+    protected function getCommunicator($uri, $client)
+    {
+        return new Communicator($uri, $client);
+    }
+
+    /**
+     * Test a simple communicator request.
+     *
+     * @return void
+     */
+    public function testSimpleRequest()
+    {
+        $client = $this->getMockClient();
+        $response = $client->send();
+        $response->expects($this->any())
```

Comment:
If you wanted to be more specific, I suspect you could use `$this->once()` instead of `$this->any()` for the two counts in this test... I don't think either of those things should be called more than once.

Reply:
True. In https://github.com/vufind-org/vufindharvest/pull/10/commits/3e8884b82ce2b859c6ed422e14cdb6692ce5c96a I updated several `expects()` calls to have a more precise frequency. This actually revealed that the test503Retry case wasn't running to completion! I turns out that `getStatusCode()` is called **twice** per request iteration (this 4 times total for the test), so an adjustment was needed there.